ios/BeaconScanner.swift

```diff
@@ -0,0 +1,225 @@
+        urls.insert("\(url)");
+      }
+      return Array(urls);
+    }
+  }
+  
+  override init() {
+    super.init()
+    
+    self.centralManager = CBCentralManager(delegate: self, queue: self.beaconOperationsQueue)
+    self.centralManager.delegate = self
+    
+  }
+  
+  func startScanning() {
+    dispatch_async(self.beaconOperationsQueue) {
+      self.startScanningSynchronized()
+    }
+  }
+  
+  func stopScanning() {
+    self.centralManager.stopScan()
+  }
+  
+  deinit {
+    self.stopScanning()
+  }
+  
+  func centralManagerDidUpdateState(central: CBCentralManager)  {
+    if central.state == CBCentralManagerState.PoweredOn && self.shouldBeScanning {
+      self.startScanning();
+    }
+  }
+  
+  func centralManager(central: CBCentralManager, didDiscoverPeripheral peripheral: CBPeripheral, advertisementData: [String : AnyObject], RSSI: NSNumber) {
+    guard let serviceData = advertisementData[CBAdvertisementDataServiceDataKey] as? [NSObject : AnyObject] else { return }
+    guard isAdvertisingURL(serviceData) else { return }
+    
+    if let url = URLForFrame(serviceData) {
+      self.addURL(url)
+    }
+  }
+  
+  private func startScanningSynchronized() {
+    dispatch_async(self.beaconOperationsQueue) {
+      if self.centralManager.state != CBCentralManagerState.PoweredOn {
+        NSLog("CentralManager state is %d, cannot start scan", self.centralManager.state.rawValue)
+        self.shouldBeScanning = true
+      } else {
+        NSLog("Starting to scan for Eddystones")
+        let services = [CBUUID(string: "FEAA"), CBUUID(string: "FED8")]
+        let options = [CBCentralManagerScanOptionAllowDuplicatesKey : true]
+        self.centralManager.scanForPeripheralsWithServices(services, options: options)
+      }
+    }
+  }
+  
+  private func addURL(url: NSURL) {
+    let count = self._urlFound.count
+    self._urlFound.insert(url)
+    if self._urlFound.count != count {
+      self.delegate?.urlContextChanged(self)
+    }
+  }
+  
+  //TODO: add support for the URL eviction
+  private func removeURL(url: NSURL) {
+    self._urlFound.remove(url)
+  }
+}
+
+private func isAdvertisingURL(advertisementFrameList: [NSObject : AnyObject]) -> Bool {
+  let eddystoneUUID = CBUUID(string: "FEAA")
+  let uribeaconUUID = CBUUID(string: "FED8")
+  let data = advertisementFrameList[uribeaconUUID]
+  if (data == nil) {
+    guard let frameData = advertisementFrameList[eddystoneUUID] as? NSData else { return false }
+    guard frameData.length > 1 else { return false }
+    
+    let count = frameData.length
+    var frameBytes = [UInt8](count: count, repeatedValue: 0)
+    frameData.getBytes(&frameBytes, length: count)
+    
+    return frameBytes[0] == 0x10
+  } else {
+    return data?.length > 0
+  }
+  
+}
+
+private func URLForFrame(advertisementFrameList: [NSObject : AnyObject]) -> NSURL? {
+  let uribeaconUUID = CBUUID(string: "FED8")
+  var data = advertisementFrameList[uribeaconUUID] as? NSData
+  if (data == nil) {
```

Comment:
IMO it would be cleaner to `return parseBeacon(data)` here if we have the `data`.